sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala

```diff
@@ -205,18 +229,164 @@ object AlluxioUtils extends Logging with Arm {
+   *
+   * returns /root
+   *
+   * @param paths the file paths in a bucket
+   * @return the root path that contains the tables
+   */
+  def getParentPathForTables(paths: Array[String]): Option[String] = {
+    val parentPathToCountMap = paths.map { e =>
+      // get table root path
+      var path = getParent(e)
+      var name = getName(path)
+      while (name.contains("=")) {
+        // it's a partition directory
+        path = getParent(path)
+        name = getName(path)
       }
+      path
+    }.map { tablePath =>
+      // get parent path of table path
+      getParent(tablePath)
+    }.map { parentPath =>
+      // add count attribute for the parent path
+      (parentPath, 1)
+    }.groupBy { case (parentPath, count) =>
+      // group by parentPath path
+      parentPath
+    }.map { case (parentPath, entries) =>
+      // sum the count
+      (parentPath, entries.size)
+    }
+
+    if (parentPathToCountMap.size > 0) {
+      // return the parentPath that contains more files
+      Some(parentPathToCountMap.maxBy { case (parentPath, count) =>
+        count
+      }._1)
+    } else {
+      None
+    }
+  }
+
+  /**
+   * Find large tables.
+   * If the avg file size for a table is bigger than the threshold, then the table is a large table.
+   *
+   * @param status    file status in a bucket
+   * @param threshold the threshold of large table
+   * @return a large table set
+   */
+  def findLargeTables(status: Array[URIStatus], threshold: Long): mutable.Set[String] = {
+    val tableSet = mutable.Set[String]()
+    // filter non-data files
+    val dataFilePaths = getDataFilePaths(status)
+
+    // find the parent path of tables
+    val parentPathOption = getParentPathForTables(dataFilePaths.map(e => e.getPath))
+    if (parentPathOption.isEmpty) {
+      return tableSet
+    }
+
+    val parentPath = parentPathOption.get
+    dataFilePaths.filter { e =>
+      e.getPath.startsWith(parentPath)
+    }.map { e =>
+      // get table name of the file, get the file size
+      val rightPart = e.getPath.substring(parentPath.size + 1)
+      val table = rightPart.substring(0, rightPart.indexOf("/"))
+      (table, e.getLength)
+    }.groupBy { case (table, size) =>
+      // group by table, one table contains multiple files
+      table
+    }.map { case (table, entries) =>
+      // calculate the avg file size for a table
+      val avgSize = entries.map(e => e._2).sum / entries.length
+      (table, avgSize)
+    }.filter { case (table, avgSize) =>
+      // filter large table
+      avgSize > threshold
+    }.foreach { case (table, avgSize) =>
+      // add the big tables to a set
+      tableSet.add(parentPath + "/" + table)
+    }
+    tableSet
+  }
+
+  def markLargeTables(bucket: String, fs: alluxio.client.file.FileSystem): Unit = {
+    // 1. list all file status, this may take some time if has lots of files.
+    val option = ListStatusPOptions.newBuilder().setRecursive(true).build()
+    val statusList = fs.listStatus(
```

Comment:
this looks very expensive to do an extra list status. This is not what I was thinking.   We already list the files in the selectedPartitions function in GpuFileSourceScanExec, which has the FileStatus with it (which has a length). We also have similar information in like createColumnarReader and  Why aren't we using that?

Reply:
Done.